src/utils/pluginDriver.ts

```diff
@@ -114,6 +115,21 @@ export function createPluginDriver(
 				warning.plugin = plugin.name || '(anonymous plugin)';
 				graph.warn(warning);
 			},
+			moduleIds: graph.moduleById.keys(),
+			getModuleInfo: (moduleId: string) => {
+				const foundModule = graph.moduleById.get(moduleId);
+				if (foundModule == null) {
+					throw new Error(`Unable to find module ${moduleId}`);
+				}
+
+				return {
+					id: foundModule.id,
+					isExternal: !!foundModule.isExternal,
+					importedIds: foundModule.isExternal
+						? []
+						: (foundModule as Module).sources.map(id => (foundModule as Module).resolvedIds[id])
```

Comment:
What happens if this is called before the module has been resolved? Is that possible?

Reply:
For now, `moduleIds` will be empty and `getModuleInfo` will throw a "module not found" error. Added a test to make sure this is the case. This may or may not be what users want but if we e.g. do not provide this information on certain hooks at all, I fear this is something that is hard to maintain properly when e.g. new hooks are added. Thus here you get a snapshot of what the current state of the graph is.

Reply:
Thanks for clarifying. It might be nice to separate the graph into two states then - resolved and unresolved, and for this function to always throw an early error when called when the graph is partially unresolved so that we aren't exposing the partial state.

Reply:
Good point but is this really an issue or more a way to avoid people relying on unreliable information? But having an internal flag marking the graph as incomplete would definitely be a future-proof way of handling this. Created #2598 to track. There may be more opinions on this issue.